Support inference with LyCORIS BOFT networks

v0xie opened the pull request:
## Description

This draft PR adds support for inference of BOFT networks trained with the dev branch of [LyCORIS](https://github.com/KohakuBlueleaf/LyCORIS). The implementation is based on @KohakuBlueleaf's implementation here: https://github.com/KohakuBlueleaf/LyCORIS/blob/dev/lycoris/modules/boft.py

Since the name "oft_diag" is the same between OFT and BOFT, I think the simplest way to differentiate the two is the dimension of the weights (OFT: 3, BOFT: 4).

The calc_updown code is slightly modified from the get_weight function in the original. I'm not sure what to do about the scale parameter so I leave it at 1.0.

## Screenshots/videos:


## Checklist:

- [x] I have read [contributing wiki page](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing)
- [x] I have performed a self-review of my own code
- [x] My code follows the [style guidelines](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing#code-style)
- [x] My code passes [tests](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Tests)

v0xie committed:
feat: support LyCORIS BOFT

v0xie committed:
fix: calculate butterfly factor

v0xie committed:
fix: add butterfly_factor fn

v0xie committed:
style: fix lint

v0xie committed:
fix: get boft params from weight shape

v0xie committed:
doc: add boft comment

v0xie committed:
Revert "fix: add butterfly_factor fn"

This reverts commit 81c16c965e532c6d86a969284c320ff8fcb0451d.

KohakuBlueleaf commented:
@v0xie For the compatibility of "load weight"

I may change the content of state dict in the future.

I think it is better to makr this pr as draft at first.

Will review and refine this in next week

KohakuBlueleaf committed:
Add rescale mechanism

LyCORIS will support save oft_blocks instead of oft_diag in the near future (for both OFT and BOFT)

But this means we need to store the rescale if user enable it.

KohakuBlueleaf committed:
Fix rescale

KohakuBlueleaf commented:
This should be fine now.

AUTOMATIC1111 commented:
There is an execution path where `is_boft` property is not added. I recommend putting 
```
self.is_kohya = False
self.is_boft = False
```
before if/else checks.

KohakuBlueleaf committed:
prevent undefined variable
